### PR TITLE
[Experimental Feature]support for common hf multimodel

### PR DESCRIPTION
**Pull request by n1ck-guo**

### CODE REVIEW

**Comment by wenhuach21** [comment]
As discussed:
1 Configure the template using a file to eliminate the need for additional coding.
2 move the code from example to autoround
3 Refine the cache code in AutoRound to improve generalizability for new models with train_batch_size=1.

**Comment by wenhuach21** [comment]
4 refine all the examples of vlm

**Comment by wenhuach21** [comment]
5 also support it in auto_round cmd

File: examples/multimodal-modeling/hf/example_template.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+    "model_type": "qwen2_vl",
+    "format_user": "<|im_start|>user\n{{content}}<|im_end|>\n",
+    "format_assistant": "<|im_start|>assistant\n{{content}}<|im_end|>\n",
+    "format_system": "<|im_start|>system\n{{content}}<|im_end|>\n",
+    "format_observation": "<|im_start|>tool\n{{content}}<|im_end|>\n<|im_start|>assistant\n",
```
Comment (wenhuach21) [inline examples/multimodal-modeling/hf/example_template.json]: create a new folder named mllm in autoround and move to it

File: auto_round/__main__.py [hunk 2]
```diff
@@ -174,7 +187,25 @@ def tune(args):
...
     low_cpu_mem_usage = False
+
+    config = AutoConfig.from_pretrained(model_name, trust_remote_code=not args.disable_trust_remote_code)
+    tokenizer = AutoTokenizer.from_pretrained(model_name, trust_remote_code=not args.disable_trust_remote_code)
+
+    model_type = config.model_type
     model_cls = AutoModel if is_glm else AutoModelForCausalLM
+
+    # for VLM
+    if args.question_path:
```
Comment (wenhuach21) [inline auto_round/__main__.py]: could not understand this args

File: auto_round/__main__.py [hunk 3]
```diff
@@ -174,7 +187,25 @@ def tune(args):
...
+        from transformers import AutoProcessor
+        if "qwen2_vl" in model_type:
+            from transformers import Qwen2VLForConditionalGeneration
+            model_cls = Qwen2VLForConditionalGeneration
+        elif "mllama" in model_type:
```
Comment (wenhuach21) [inline auto_round/__main__.py]: this should be in other place to support API usage,

File: auto_round/__main__.py [hunk 4]
```diff
@@ -283,16 +311,31 @@ def tune(args):
...
+    from auto_round.vlm_dataset import get_dataloader
+    if args.template:
+        template = args.template
+    else:
+        template = model_type
+    dataloader = get_dataloader(template, tokenizer, question_path=args.question_path, image_path=args.image_path, seqlen=args.seqlen)
+
+    if args.question_path:
+        from auto_round.utils import get_multimodal_block_names
+        dataset = dataloader
```
Comment (wenhuach21) [inline auto_round/__main__.py]: same issue, we should support these models in both command and API usage

File: auto_round/autoround.py [hunk 5]
```diff
@@ -1623,6 +1623,7 @@ def __init__(
         )
 
 
+
 class AutoMLLMROund(AutoRound):
     def __init__(
```
Comment (wenhuach21) [inline auto_round/autoround.py]: AutoRoundMLLM, O->o

File: auto_round/__main__.py [hunk 6]
```diff
@@ -154,6 +154,16 @@ def setup_parser():
     parser.add_argument("--fp_layers_list", default="", type=str,
                         help="List of Layers to maintain original data type")
 
+    # ======================= VLM =======================
+    parser.add_argument("--quant_vision", action='store_true',
+                        help="To determine whether the quantization should handle vision component.")
+
+    parser.add_argument("--images", default="coco", type=str,
```
Comment (wenhuach21) [inline auto_round/__main__.py]: replace this arg with others 
All we need is a dataset name The string can represent either a Hugging Face dataset name or a local file path.

The dataset metadata file must be in JSON format and should include information about text and other modalities, such as images and videos.

Users have the option to provide a dictionary or a string(only for only one extra modality)supply additional information, enabling the retrieval of corresponding images, videos, or audio files.

File: auto_round/autoround.py [hunk 7]
```diff
@@ -1622,3 +1624,94 @@ def __init__(
...
+class AutoMLLMROund(AutoRound):
+    def __init__(
+            self,
+            model,
+            tokenizer,
+            bits: int = 4,
+            group_size: int = 128,
+            sym: bool = False,
+            layer_config: dict = None,
+            enable_full_range: bool = False,  ##for symmetric, TODO support later
```
Comment (wenhuach21) [inline auto_round/autoround.py]: this have been removed in the main branch

File: examples/multimodal-modeling/hf/main.py [hunk 8]
```diff
@@ -112,10 +112,13 @@ def setup_args():
...
-                        help="The dataset for quantization training. It can be a custom one.")
+    parser.add_argument("--extra_data_dir", default="", type=str,
+                        help="Dataset dir for storing images/audio/videos. "
+                        "Can be a dir path or multiple dir path with format as "
+                        "'image=path_to_image,video=path_to_video,audio=path_to_audio'"
+                        "By default, it will search in the relative path.")
     
-    parser.add_argument("--question_path", default=None, type=str,
+    parser.add_argument("--dataset", default=None, type=str,
                             help="The dataset for quantization training. It can be a custom one.")
```
Comment (wenhuach21) [inline examples/multimodal-modeling/hf/main.py]: if __main__  could support all the functions in the example, we could delete the example

File: auto_round/mllm/mllm_dataset.py [hunk 9]
```diff
@@ -129,10 +129,23 @@ def vlm_encode(sources, template: "Template"):
...
+        result = {}
+        dir_path = dir_path.split(",")
+        for _path in dir_path:
+            k, v = _path.split('=')
+            if k in ['image', 'video', 'audio']:
+                result[k] = v
+        return result
+            
+
 class LlavaDataset(Dataset):
```
Comment (wenhuach21) [inline auto_round/mllm/mllm_dataset.py:48]: support register, we may use other dataset. Also support customized dataset for users which allow users to pass a json which includes text or chat text and image path as the dataset

File: auto_round/__main__.py [hunk 10]
```diff
@@ -365,15 +383,116 @@ def eval(args):
     print(make_table(res))
 
 
+def tune_mllm(args):
+    if args.act_bits <= 8:
+        print(
+            "Warning, activation quantization is an experiment feature")
+    
+    if "marlin" in args.format and args.sym == False:
+        assert False, "marlin backend only supports sym quantization, please set --sym"
```
Comment (wenhuach21) [inline auto_round/__main__.py]: the --sym has been replaced by --asym,  please help fix this,  also help check main.py

File: auto_round/mllm/templates/cogvlm2.json [hunk 11]
```diff
@@ -0,0 +1,6 @@
+{
+    "model_type": "cogvlm2",
+    "format_user": "Question: {{content}} ",
+    "format_assistant": "Answer: {{content}}\n",
+    "plugin": "cogvlm2"
```
Comment (wenhuach21) [inline auto_round/mllm/templates/cogvlm2.json]: For the plugin, is it possible to support a new model without adding additional code? If that's not feasible, then using JSON might not be the best approach.

Comment (n1ck-guo) [reply auto_round/mllm/templates/cogvlm2.json]: We have default plugin and some pre-defined plugins to choose. Also, if user want to use customized plugin, they can use register func to add a new one.

File: auto_round/__main__.py [hunk 12]
```diff
@@ -388,9 +388,6 @@ def tune_mllm(args):
         print(
             "Warning, activation quantization is an experiment feature")
     
-    if "marlin" in args.format and args.sym == False:
-        assert False, "marlin backend only supports sym quantization, please set --sym"
```
Comment (wenhuach21) [inline auto_round/__main__.py]: why delete, should be disable --asym

File: auto_round/autoround.py [hunk 13]
```diff
@@ -1622,3 +1626,195 @@ def __init__(
...
+            sym: bool = False,
+            layer_config: dict = None,
+            batch_size: int = 8,
+            amp: bool = True,
+            device: str = None,
+            lr_scheduler=None,
+            dataset: Union[str, list, tuple, torch.utils.data.DataLoader] = None,
+            extra_data_dir: Union[str, torch.utils.data.DataLoader] = None,
+            template: Union[str, Template] = None,
+            quant_vision: bool = False,
```
Comment (wenhuach21) [inline auto_round/autoround.py]: better change a name, quant_nontext_module or something like that to extend to other modality later

File: auto_round/__main__.py [hunk 14]
```diff
@@ -365,15 +383,116 @@ def eval(args):
...
+    if "hpu" in device_str:
+        torch_dtype = torch.bfloat16
+    
+    torch.manual_seed(1234)
+
+    # load_model
+    from auto_round.mllm import load_mllm
+    model, tokenizer, processor = load_mllm(model_name, device_map="auto", torch_dtype=torch_dtype, trust_remote_code=not args.disable_trust_remote_code)
+
+    from auto_round import AutoMLLMRound
```
Comment (wenhuach21) [inline auto_round/__main__.py]: change all the names to use AutoRoundXXX?

File: auto_round/autoround.py [hunk 15]
```diff
@@ -1622,3 +1626,195 @@ def __init__(
...
+            act_dynamic=act_dynamic,
+            enable_norm_bias_tuning=enable_norm_bias_tuning,
+            quant_block_list=quant_block_list,
+            **kwargs,
+        )
+        
+
+    def calib(self, nsamples, bs):
+        if isinstance(self.dataset, str):
+            dataset = self.dataset.replace(" ", "")
```
Comment (wenhuach21) [inline auto_round/autoround.py]: move this class to mllm?

File: auto_round/mllm/mllm_dataset.py [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+            model_type_or_template,
+            model,
+            tokenzier,
+            dataset_path,
+            extra_data_dir,
+            max_length,
+            padding=True,
+            truncation=True,
+            ) -> None:
+        super().__init__()
```
Comment (wenhuach21) [inline auto_round/mllm/mllm_dataset.py:62]: Please enhance the code by adding some comments and better add a README to include instructions on how to prepare llava dataset , support new models, etc. in this folder. Additionally, better test the code with an unseen model to evaluate its generalization capabilities.

File: auto_round/mllm/__init__.py [hunk 17]
```diff
@@ -15,4 +15,4 @@
 from .mllm_dataset import get_mllm_dataloader
 from .template import Template, get_template, TEMPLATES
 from .utils import load_mllm
-from .autoround import AutoRoundMLLM
\ No newline at end of file
+from .mllm_autoround import AutoRoundMLLM
```
Comment (wenhuach21) [inline auto_round/mllm/__init__.py]: autoround_mmlm, follow our name style?

File: auto_round/mllm/utils.py [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+    if "qwen2_vl" in model_type:
+        from transformers import Qwen2VLForConditionalGeneration
+        model = Qwen2VLForConditionalGeneration.from_pretrained(
+            pretrained_model_name_or_path, trust_remote_code=trust_remote_code, **kwargs)
+    elif "mllama" in model_type:
+        from transformers import MllamaForConditionalGeneration
+        model = MllamaForConditionalGeneration.from_pretrained(
+            pretrained_model_name_or_path, trust_remote_code=trust_remote_code, attn_implementation="eager", **kwargs)
+    else:
+        model = AutoModelForCausalLM.from_pretrained(
```
Comment (wenhuach21) [inline auto_round/mllm/utils.py]: is there an AutoModeForConditionalGeneraion API in HF?

Comment (n1ck-guo) [reply auto_round/mllm/utils.py]: There is no such api

File: auto_round/mllm/utils.py [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+from transformers import AutoModelForCausalLM, AutoProcessor, AutoTokenizer, AutoConfig
+
+
+def load_mllm(pretrained_model_name_or_path, **kwargs):
+    """Load MLLMs.
+    Args:
+        pretrained_model_name_or_path (str): The name or path of the pretrained model.
+        kwargs: Variable number of keyword arguments.
+    """
+    trust_remote_code = True
```
Comment (wenhuach21) [inline auto_round/mllm/utils.py]: I think we should delete this function and just use the inner side code. This function lacks generality and robustness, which may confuse users when attempting to call it

File: auto_round/mllm/utils.py (hunk 18 above)
Comment (wenhuach21) [reply auto_round/mllm/utils.py]: then better raise error

File: auto_round/__main__.py [hunk 20]
```diff
@@ -459,8 +459,25 @@ def tune_mllm(args):
...
+    tokenizer.processor = processor
+    model_type = config.model_type
+    if "qwen2_vl" in model_type:
+        from transformers import Qwen2VLForConditionalGeneration
+        model = Qwen2VLForConditionalGeneration.from_pretrained(
+            model_name, trust_remote_code=not args.disable_trust_remote_code)
+    elif "mllama" in model_type:
+        from transformers import MllamaForConditionalGeneration
+        model = MllamaForConditionalGeneration.from_pretrained(
+            model_name, trust_remote_code=not args.disable_trust_remote_code, attn_implementation="eager")
```
Comment (wenhuach21) [inline auto_round/__main__.py]: remove eager later

File: auto_round/mllm/README.md [hunk 21]
```diff
@@ -1,11 +1,15 @@
...
 from auto_round import AutoRoundMLLM
+from transformers import Qwen2VLForConditionalGeneration, AutoProcessor, AutoTokenizer
 
 model_name = "Qwen/Qwen2-VL-2B-Instruct"
-model, tokenizer, processor = load_mllm(model_name, device_map="auto")
+tokenizer = AutoTokenizer.from_pretrained(model_name)
+processor = AutoProcessor.from_pretrained(model_name, trust_remote_code=trust_remote_code)
+tokenizer.processor = processor
+model = Qwen2VLForConditionalGeneration.from_pretrained(
+    model_name, trust_remote_code=True, device_map="auto")
```
Comment (wenhuach21) [inline auto_round/mllm/README.md]: don't use auto,

**Review by wenhuach21 — APPROVED** [review]

**Pull request merged**